[FLINK-21445][kubernetes] Application mode does not set the configuration when building PackagedProgram

SteNicholas opened the pull request:
## What is the purpose of the change

*Application mode uses `ClassPathPackagedProgramRetriever` to create the `PackagedProgram` which doesn't set the configuration. This causes some client configurations not take effect. `ClassPathPackagedProgramRetriever` should refactor to have separate classes for the Python-based PackagedProgram and the PackagedProgram with and without a jar.*

## Brief change log

  - *`ClassPathPackagedProgramRetriever` refactors to have separate implementations which are `PythonBasedPackagedProgramRetriever`, `JarFilePackagedProgramRetriever` and `ScanClassPathPackagedProgramRetriever`, which set the configuration to `PackagedProgram`.*
  - *`ApplicationClusterEntryPoint` gets the `PackagedProgramRetriever` with setting the configuration.*

## Verifying this change

  - *`ClassPathPackagedProgramRetrieverTest` adds the test case `testGetPackagedProgramWithConfiguration` to verify whether get the `PackagedProgram` with the specified configuration.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

SteNicholas commented:
@XComp , could you please help to review this pull request?

wangyang0918 commented:
@SteNicholas Thanks for creating this PR. I have gone though the implementation carefully and want to share some thoughts. Instead of inheritance here, I would suggest to use the composition. Maybe we could refactor this component by the following steps.

* Introduce the `PackagedProgramRetrieverAdapter`(maybe other better name), which will provide a builder and select a proper `PackagedProgramRetriever` based on the specified properties. It is just like what you have done in the `ClassPathPackagedProgramRetriever#Builder`
* Make `JarFilePackagedProgramRetriever`, `PythonBasedPackagedProgramRetriever`, `ScanClassPathPackagedProgramRetriever`(could be renamed to the new `ClassPathPackagedProgramRetriever`) directly implement the `PackagedProgramRetriever` interface
* Maybe we could add the `AbstractPackagedProgramRetriever` if the above three cases have some common member variables or methods to share.

After this refactor, we also need to split the tests in `ClassPathPackagedProgramRetrieverTest` to separate files. If the three cases have some common tests, then they could be implemented in the `PackagedProgramRetrieverTestBase`. At last, we should not forget to add some dedicated tests for the `PackagedProgramRetrieverAdapter`.

XComp commented:
Sorry for not getting back earlier: I second @wangyang0918 proposal. The code should be easier to test after the proposed change. Additionally, the `JarsOnClassPath` singleton seems to be only relevant to what is now `ScanClassPathPackagedProgramRetriever`.

SteNicholas commented:
@wangyang0918 , @XComp , thanks for detailed review. I have followed the comments of @wangyang0918 , and taken additional remark of @XComp . Please help to review again if you are available.

SteNicholas commented:
@wangyang0918 , @XComp , thanks for your detailed review. I have followed the comments of @wangyang0918 and taken the additional remark @XComp . Please help to review this pull request again.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/ClassPathPackagedProgramRetrieverTest.java:
```diff
@@ -285,25 +148,55 @@ public void testJobGraphRetrievalFailIfDoesNotFindTheEntryClassInTheJobDir()
 
     @Test
-    public void testRetrieveCorrectUserClasspathsWithoutSpecifiedEntryClass()
+    public void testSavepointRestoreSettings()
```
This should also be tested for `JarFilePackagedProgramRetriever` and `PythonBasedPackagedProgramRetriever`. Maybe we could add it to the `PackagedProgramRetrieverTestBase`.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/ClassPathPackagedProgramRetrieverTest.java:
```diff
@@ -312,87 +205,64 @@ public void testRetrieveCorrectUserClasspathsWithSpecifiedEntryClass()
-                        .setJarFile(testJar)
-                        .build();
-        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+    public void testGetPackagedProgramWithConfiguration() throws IOException, FlinkException {
```
This should also be tested for JarFilePackagedProgramRetriever and PythonBasedPackagedProgramRetriever.

wangyang0918 reviewed (changes requested):
Thanks for updating this PR. It looks really good to me now. I left some minor comments. Please have a look.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/ClassPathPackagedProgramRetrieverTest.java:
```diff
@@ -285,25 +148,55 @@ public void testJobGraphRetrievalFailIfDoesNotFindTheEntryClassInTheJobDir()
 
     @Test
-    public void testRetrieveCorrectUserClasspathsWithSpecifiedEntryClass()
+    public void testRetrieveCorrectUserClassPathsWithSpecifiedEntryClass()
```
Correct me if I'm wrong here but don't we want to move this into `PackaedProgramRetrieverTestBase` as well?

tillrohrmann commented:
What's the state of this PR @XComp?

SteNicholas commented:
@tillrohrmann , I will update this pull request through following the comments above.

XComp commented:
Ok, thanks @SteNicholas . I'll give it another pass when you're done.

XComp commented:
Any updates from your side, @SteNicholas ?

XComp commented:
@SteNicholas Did you have a chance to work on it? Or do you have an idea when to proceed?

SteNicholas commented:
> @SteNicholas Did you have a chance to work on it? Or do you have an idea when to proceed?

@XComp , I’m adding the `PythonBasedPackagedProgramRetrieverTest`. And I will update this pull request today.

SteNicholas commented:
@XComp @wangyang0918 I have followed your detailed review comments. Please help to review the update of the pull request.

SteNicholas commented:
@XComp @wangyang0918  I have followed the detailed review comments. Please help to review the update of the pull request.

XComp reviewed (changes requested):
Hi @SteNicholas , thanks for your efforts. Thanks for your patience. I gave it another pass and am still puzzled by the complexity of the actual code. I digged a bit deeper into what the code actually does (I should have done that earlier already).

In the end, here are a few of my findings which we might want to address in the refactoring (please validate the correctness):
* `PythonBasedPackagedProgramRetriever` is just a `JarFilePackagedProgramRetriever` with a specific `jarFile` and `jobClassName`. Hence, we do not necessarily need a dedicated class for it but pass in these specific parameter values.
* The purpose of the different subclasses of `PackagedProgramRetriever` is actually only to select the right `jar` and `jobClassName`. Hence, we could create a generic `PackagedProgramRetriever` class that takes all the common parameters and expects some kind of interface (e.g. `EntryClassInformationProvider`) providing `getJarFile()` and `getJobClassName()` as methods. We could then provide a generic `FromJarEntryClassInformationProvider` which substitutes the `JarFilePackagedProgramRetriever`. This can be also used for the Python-based version. Another implementation would look for the user/system class path as it's done in `ClasspathPackagedProgramRetriever`. This approach enables us to properly unit test the individual components and avoid redundant code. What do you think?
* Another finding: It looks like `JarsOnClasspath` is the only jar provider that is used in production? @wangyang0918 do you have an idea whether that's also a missing option? If not, we could consider removing the configurability and make it static to simplify the code even more.

I'm looking forward to your opinions on that, @wangyang0918 & @SteNicholas . Let me know if I need to clarify things about my proposal.

SteNicholas commented:
@XComp , thanks for your detailed comments for the refactoring.  Based on your finding, I thought that:

1. `PythonBasedPackagedProgramRetriever` is indeed a `JarFilePackagedProgramRetriever` with a specific `jarFile` and `jobClassName`, which could be removed and pass in these specific parameter values in `JarFilePackagedProgramRetriever`.
2. `PackagedProgramRetriever` implementation could be similified by introducing interface `EntryClassInformationProvider` providing `getJarFile()` and `getJobClassName()` as methods.

wangyang0918 commented:
@XComp IIUC, what you are suggesting is using the composition instead of inheritance. I agree with you that it is easier to test and could make the contract clearer.

For `JarsOnClasspath`, it could be subsumed in the `EntryClassInformationProvider` implementation for classpath.

XComp commented:
@SteNicholas just a kind ping. :-) do you have an idea when you could continue working on this issue?

SteNicholas commented:
> @SteNicholas just a kind ping. :-) do you have an idea when you could continue working on this issue?

Of course, I'm addressing your comments. Sorry for later update for this request.

XComp commented:
Hi @SteNicholas , how is the this issue's progress? I'm wondering whether you could give a time estimate on when you're able to close it. I understand that there are other things that need to be done on your side as well. I could take over if that helps.

XComp commented:
Hi @SteNicholas , I'm going to pick up that one if that's ok with you. ...just to have it off the list.

SteNicholas commented:
> Hi @SteNicholas , I'm going to pick up that one if that's ok with you. ...just to have it off the list.

OK.
